src/hotspot/share/opto/vectornode.cpp

```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
   return this;
 }
 
+Node* SelectFromTwoVectorNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  int num_elem = vect_type()->length();
+  BasicType elem_bt = vect_type()->element_basic_type();
+  if (Matcher::match_rule_supported_vector(Op_SelectFromTwoVector, num_elem, elem_bt)) {
+    return nullptr;
+  }
+  Node* index_vec = in(1);
+  Node* src1  = in(2);
+  Node* src2  = in(3);
+
+  // Lower the IR to constituents operations.
+  //   SelectFromTwoVectorNode =
+  //     (VectorBlend
+  //         (VectorRearrange SRC1, INDEX)
+  //         (VectorRearrange SRC2, NORM_INDEX)
+  //         MASK)
+  // This shall prevent an intrinsification failure and associated argument
+  // boxing penalties.
+
+  auto lane_count_type = [&]() {
+    switch(elem_bt) {
+      case T_BYTE:
+      case T_SHORT:
+      case T_INT:
+      case T_FLOAT:
+        return static_cast<const Type*>(TypeInt::make(num_elem));
+      case T_DOUBLE:
+      case T_LONG:
+        return static_cast<const Type*>(TypeLong::make(num_elem));
+      default:
+        fatal("Unsupported vectortype (%s)", type2name(elem_bt));
+        return static_cast<const Type*>(nullptr);
+    }
+  };
+
+  auto make_integral_index_vec = [&](Node* index_vec) {
+    switch(elem_bt) {
+      case T_FLOAT:
+        return phase->transform(new VectorCastF2XNode(index_vec, TypeVect::make(T_INT, num_elem)));
+      break;
+      case T_DOUBLE:
+        return phase->transform(new VectorCastD2XNode(index_vec, TypeVect::make(T_LONG, num_elem)));
+      break;
+      default:
+        return index_vec;
+    }
+  };
+
+  auto get_integal_type = [&](BasicType elem_bt) {
+    switch(elem_bt) {
+      case T_FLOAT:  return T_INT;
+      case T_DOUBLE: return T_LONG;
+      default: return elem_bt;
+    }
+  };
```

Comment:
This is definitely a style question. But it might be nice to make these functions member functions. They now kinda disrupt the flow of the `::Ideal` method. And in some cases you use the captured variables, and in other cases you pass them in explicitly, even though they already exist in the captured scope... consistency would be nice.